### PR TITLE
Makes the colon catcher deterministic

### PR DESCRIPTION
**Pull request by Cyberboss**
- line_num now increments during multiline strings
- line increments now reset a number of things
- ternary_on_line var repurposed to an int to count the number of ternarys
- in_string is now false while in_embedded_statement

- [ ] Ignores text colons in multi line strings
- [x] Ignores loop labels
- [ ] Ignores multi line ternarys (god why do these exist)

Result: no more questionable colons due to ternarys. Updated the operator log

### CODE REVIEW

File: tools/ColonCatcher/2017-08-09-colon_operator_log.txt [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+..code\modules\mob\living\simple_animal\slime\life.dm Lines: 409, 417, 419 Total Colons: 3
+..code\modules\ninja\suit\suit_initialisation.dm Lines: 73 Total Colons: 1
+..code\modules\paperwork\photography.dm Lines: 225, 226 Total Colons: 2
+..code\modules\power\smes.dm Lines: 52 Total Colons: 1
+..code\modules\reagents\chemistry\reagents\food_reagents.dm Lines: 393 Total Colons: 2
+..code\__HELPERS\icons.dm Lines: 715, 752, 757, 762, 763, 764, 765, 774, 786, 788 Total Colons: 14
+..code\__HELPERS\mobs.dm Lines: 352 Total Colons: 1
+..code\__HELPERS\_lists.dm Lines: 323, 471 Total Colons: 2
+..code\__HELPERS\sorts\__main.dm Lines: 391, 491 Total Colons: 2
+52/2191 files have colons in them
```
Comment (RemieRichards) [inline tools/ColonCatcher/2017-08-09-colon_operator_log.txt]: almost 500 new files and yet 350 less colons.
we're truly doing gods work.

File: tools/ColonCatcher/ColonCatcher.py [hunk 2]
```diff
@@ -58,17 +58,18 @@ def scan_dm_file(_file):
...
         for char in characters:
             #Line info
-            if char == "\n" or char == "\r":
+            if char == "\n":
+                line_num += 1
                 if not in_string:
-                    ternary_on_line = False #Stop any old ternary operation
-                    line_num += 1
                     in_embed_statement = 0
+                    in_string = False
```
Comment (RemieRichards) [inline tools/ColonCatcher/ColonCatcher.py]: but mutliline strings though...

```
var/mstring = {" I
am
not
happy
with
you"}
```

File: tools/ColonCatcher/ColonCatcher.py [hunk 3]
```diff
@@ -82,6 +83,14 @@ def scan_dm_file(_file):
                     if char == "\"":
                         in_string = True
 
+                    #ternary statements, True when not in_string
+                    if char == "?":
```
Comment (RemieRichards) [inline tools/ColonCatcher/ColonCatcher.py]: will need updating with 512's ```?.``` operator

File: tools/ColonCatcher/ColonCatcher.py [hunk 4]
```diff
@@ -91,39 +100,20 @@ def scan_dm_file(_file):
...
-                    if not in_string:
+                if (char == ":") and (not in_string):
+                    if ternarys_on_line > 0:
+                        ternarys_on_line -= 1
+                    else:
                         colon_count += 1
                         data = str(line_num)
-                        if ternary_on_line:
-                            data += "?"
                         if not data in lines_with_colons: #only add the line twice if it's like: 76, 76? (eg: a "bad" colon and a ternary colon)
```
Comment (RemieRichards) [inline tools/ColonCatcher/ColonCatcher.py]: ? lines can't exist anymore, so update this logic/the comment

File: tools/ColonCatcher/ColonCatcher.py (hunk 2 above)
Comment (RemieRichards) [inline tools/ColonCatcher/ColonCatcher.py]: but mutliline strings though...

```
var/mstring = {"I
am
not
happy
with
you"}
```

File: tools/ColonCatcher/ColonCatcher.py (hunk 3 above)
Comment (RemieRichards) [inline tools/ColonCatcher/ColonCatcher.py]: will need updating with 512's ```?.``` operator

File: tools/ColonCatcher/ColonCatcher.py (hunk 4 above)
Comment (RemieRichards) [inline tools/ColonCatcher/ColonCatcher.py]: ? lines can't exist anymore, so update this logic/the comment

**Review by RemieRichards — CHANGES_REQUESTED** [review]
Looks good, just a few points

**Comment by RemieRichards** [comment]
also where was it not deterministic before?
If you mean in the logic of correctly detecting colons that's not determinism that's just fixing it being wonky.

File: tools/ColonCatcher/2017-08-09-colon_operator_log.txt [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+..code\game\objects\items\devices\PDA\cart.dm Lines: 263, 393, 410, 701, 702, 703, 706, 707, 708 Total Colons: 9
+..code\modules\admin\admin.dm Lines: 41, 137, 144, 156, 261, 307, 321, 367 Total Colons: 9
+..code\modules\admin\sql_message_system.dm Lines: 268 Total Colons: 2
+..code\modules\admin\stickyban.dm Lines: 188 Total Colons: 1
+..code\modules\admin\DB_ban\functions.dm Lines: 87, 88, 89, 315, 316, 317 Total Colons: 6
+..code\modules\admin\verbs\adminjump.dm Lines: 91 Total Colons: 1
+..code\modules\admin\verbs\debug.dm Lines: 321, 338, 464, 467 Total Colons: 4
+..code\modules\admin\verbs\diagnostics.dm Lines: 82 Total Colons: 4
+..code\modules\admin\verbs\randomverbs.dm Lines: 847, 852, 858, 864, 870, 876, 882, 888, 894, 900, 906, 912, 918, 924, 930, 936, 942, 948 Total Colons: 18
+..code\modules\admin\verbs\SDQL2\SDQL_2.dm Lines: 392, 434 Total Colons: 2
```
Comment (RemieRichards) [inline tools/ColonCatcher/2017-08-09-colon_operator_log.txt]: both of these are in strings
https://github.com/tgstation/tgstation/blob/master/code/modules/admin/verbs/SDQL2/SDQL_2.dm#L392
https://github.com/tgstation/tgstation/blob/master/code/modules/admin/verbs/SDQL2/SDQL_2.dm#L434

Not bothered to check other files, this is enough to prove it's still a bit wonk.

File: tools/ColonCatcher/2017-08-09-colon_operator_log.txt [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+..code\game\gamemodes\objective.dm Lines: 847 Total Colons: 1
+..code\game\machinery\newscaster.dm Lines: 364, 415, 428, 478, 944 Total Colons: 5
+..code\game\machinery\overview.dm Lines: 99, 245 Total Colons: 2
+..code\game\machinery\robot_fabricator.dm Lines: 21, 23, 24, 25, 28 Total Colons: 5
+..code\game\machinery\status_display.dm Lines: 153 Total Colons: 4
+..code\game\machinery\computer\communications.dm Lines: 504, 635 Total Colons: 2
+..code\game\machinery\computer\medical.dm Lines: 62, 67 Total Colons: 2
+..code\game\machinery\computer\security.dm Lines: 121, 128, 209, 229, 385, 401 Total Colons: 6
+..code\game\machinery\doors\brigdoors.dm Lines: 193 Total Colons: 4
+..code\game\mecha\mecha.dm Lines: 717 Total Colons: 1
```
Comment (RemieRichards) [inline tools/ColonCatcher/2017-08-09-colon_operator_log.txt]: https://github.com/tgstation/tgstation/blob/master/code/game/mecha/mecha.dm#L717

Also highlighting this one actually, as it's a multi line ternary (DUN DUN DUN!!!1!!)

Comment (Cyberboss) [inline tools/ColonCatcher/2017-08-09-colon_operator_log.txt]: WHY!

**Comment by Cyberboss** [comment]
By not deterministic I meant not being able o detect embedded ternarys

**Comment by RemieRichards** [comment]
Then you don't know what determinism means :P

**Comment by Cyberboss** [comment]
lazy

**Pull request closed without merging**